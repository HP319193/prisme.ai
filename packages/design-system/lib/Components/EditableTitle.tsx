import { useState } from 'react';
import { Typography } from 'antd';

const { Title } = Typography;

export interface EditableTitleProps {
  value: string;
  onChange: (value: string) => void;
  level?: 1 | 2 | 3 | 4 | 5;
  className?: string;
}

const EditableTitle = ({ value, onChange, ...props }: EditableTitleProps) => {
  return (
    <Title editable={{ onChange }} {...props}>
      {value}
    </Title>
  );
};

export default EditableTitle;
