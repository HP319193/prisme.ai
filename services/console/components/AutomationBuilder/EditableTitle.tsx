import { useContentEditable } from '@prisme.ai/design-system';
import { forwardRef, HTMLAttributes, useEffect, useRef, useState } from 'react';
import LocalizedInput from '../LocalizedInput';

interface EditableTitleProps
  extends Omit<HTMLAttributes<HTMLInputElement>, 'onChange'> {
  value: Prismeai.LocalizedText;
  onChange: (v: Prismeai.LocalizedText) => void;
  onEnter?: (v: Prismeai.LocalizedText) => void;
}

const Input = forwardRef<any, any>(function Input(
  { value, onChange, onBlur, onEnter, ...props },
  ref
) {
  const contentEditable = useContentEditable({
    onInput: (v) => onChange({ target: { value: v } }),
    onEnter,
  });

  return (
    <span
      {...contentEditable}
      onBlur={(e) => {
        e.preventDefault();
        contentEditable.onBlur(e);
      }}
      onKeyDown={(e) => {
        contentEditable.onKeyDown(e);
      }}
      ref={ref}
      {...props}
      className={`min-w-[5rem] max-w-[50vw] text-ellipsis overflow-hidden mr-8 mt-[0.4rem] ${props.className}`}
    >
      {value}
    </span>
  );
});

export const EditableTitle = ({
  value,
  onChange,
  onEnter,
}: EditableTitleProps) => {
  return (
    <LocalizedInput
      value={value}
      onChange={onChange}
      Input={Input}
      InputProps={{ onEnter }}
    />
  );
};

export default EditableTitle;
