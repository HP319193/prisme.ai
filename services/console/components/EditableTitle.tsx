import { FC, LegacyRef, useCallback, useRef, useState } from 'react';
import { Inplace, InplaceDisplay, InplaceContent } from 'primereact/inplace';
import { InputText } from 'primereact/inputtext';
interface EditableTitleProps {
  title: string;
  onChange: (v: string) => void;
}
export const EditableTitle: FC<EditableTitleProps> = ({ title, onChange }) => {
  const [value, setValue] = useState(title);
  const inplace = useRef<Inplace & { close: Function }>(null);

  const submit = useCallback(() => {
    if (value.length !== 0) {
      onChange(value);
    }
    inplace.current && inplace.current.close();
  }, [onChange, value]);
  return (
    <Inplace closable className="flex" ref={inplace}>
      <InplaceDisplay>
        <div className="flex">{title}</div>
      </InplaceDisplay>
      <InplaceContent>
        <InputText
          value={value}
          onChange={({ target: { value } }) => setValue(value)}
          autoFocus
          onKeyDown={({ key }) => key === 'Enter' && submit()}
          onBlur={submit}
        />
      </InplaceContent>
    </Inplace>
  );
};
export default EditableTitle;
