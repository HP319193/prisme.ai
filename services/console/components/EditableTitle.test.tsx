import EditableTitle from './EditableTitle';
import renderer, { act } from 'react-test-renderer';
import { InputText } from 'primereact/inputtext';
// @ts-ignore because its a mock
import { close } from 'primereact/inplace';

jest.mock('primereact/inplace', () => {
  const EmptyComponent = require('../__mocks__/EmptyComponent').default;

  const close = jest.fn();
  class Inplace extends EmptyComponent {
    close = close;
  }
  return {
    Inplace,
    InplaceDisplay: EmptyComponent,
    InplaceContent: EmptyComponent,
    close,
  };
});

it('should render', () => {
  const onChange = jest.fn();
  const root = renderer.create(
    <EditableTitle title="foo" onChange={onChange} />
  );
  expect(root.toJSON()).toMatchSnapshot();
});

it('should change value on Enter', () => {
  const onChange = jest.fn();
  const root = renderer.create(
    <EditableTitle title="foo" onChange={onChange} />
  );
  act(() => {
    root.root
      .findByType(InputText)
      .props.onChange({ target: { value: 'foo' } });
    root.root.findByType(InputText).props.onKeyDown({ key: 'Enter' });
  });
  expect(onChange).toHaveBeenCalledWith('foo');
  expect(close).toHaveBeenCalled();
});

it('should change value on Blur', () => {
  const onChange = jest.fn();
  const root = renderer.create(
    <EditableTitle title="foo" onChange={onChange} />
  );
  act(() => {
    root.root
      .findByType(InputText)
      .props.onChange({ target: { value: 'foo' } });
    root.root.findByType(InputText).props.onBlur();
  });
  expect(onChange).toHaveBeenCalledWith('foo');
  expect(close).toHaveBeenCalled();
});

it('should not change value if empty', () => {
  const onChange = jest.fn();
  const root = renderer.create(
    <EditableTitle title="foo" onChange={onChange} />
  );
  act(() => {
    root.root.findByType(InputText).props.onChange({ target: { value: '' } });
  });
  act(() => {
    root.root.findByType(InputText).props.onBlur();
  });
  expect(onChange).not.toHaveBeenCalled();
});
