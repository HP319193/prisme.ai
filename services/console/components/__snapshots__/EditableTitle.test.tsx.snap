// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
Array [
  <div
    className="flex"
  >
    foo
  </div>,
  <input
    autoFocus={true}
    className="p-inputtext p-component p-filled"
    onBlur={[Function]}
    onChange={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    value="foo"
  />,
]
`;
